panic: send on closed channel

@inge4pres, I was running some tests with a master build locally and ran into the following error:

```
panic: send on closed channel

goroutine 41 [running]:
github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProjectsWith.func1(0xc000192480, 0xc0000f61e0, 0xc00012fd00, 0xc000192780)
	/go/gitlab-ci-pipelines-exporter/cmd/gitlab.go:483 +0x1b2
created by github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProjectsWith
	/go/gitlab-ci-pipelines-exporter/cmd/gitlab.go:476 +0x120
```

I'll try to dig into it later on, just to let you know! :)

nb: probably related to https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/pull/88

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@ vendor
 **/.DS_Store
 gitlab-ci-pipelines-exporter
 coverage.out
+.test

--- a/Makefile
+++ b/Makefile
@@ -45,7 +45,7 @@ misspell: setup ## Test code with misspell
 
 .PHONY: test
 test: ## Run the tests against the codebase
-	go test -v ./...
+	go test -v -race ./...
 
 .PHONY: install
 install: ## Build and install locally the binary (dev purpose)

--- a/cmd/gitlab.go
+++ b/cmd/gitlab.go
@@ -5,6 +5,7 @@ import (
 	"regexp"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/google/go-cmp/cmp"
@@ -444,9 +445,11 @@ func (c *Client) discoverWildcards() {
 
 func (c *Client) pollWithWorkersUntil(stop <-chan struct{}) {
 	log.Infof("%d project(s) configured for polling", len(cfg.Projects))
-	pollErrors := c.pollProjectsWith(cfg.MaximumProjectsPollingWorkers, stop, cfg.Projects...)
-	for _, err := range pollErrors {
-		log.Errorf("%v", err)
+	pollErrors := c.pollProjectsWith(cfg.MaximumProjectsPollingWorkers, c.pollProject, stop, cfg.Projects...)
+	for err := range pollErrors {
+		if err != nil {
+			log.Errorf("%v", err)
+		}
 	}
 }
 
@@ -466,40 +469,39 @@ func (c *Client) findProjectsFromWildcards() error {
 	return nil
 }
 
-func (c *Client) pollProjectsWith(numWorkers int, until <-chan struct{}, projects ...Project) []error {
-	var errs []error
+func (c *Client) pollProjectsWith(numWorkers int, doing func(Project) error, until <-chan struct{}, on ...Project) <-chan error {
 	errorStream := make(chan error)
-	defer close(errorStream)
-	projectsToPoll := make(chan Project, len(projects))
+	projectsToPoll := make(chan Project, len(on))
+	// sync closing the error channel via a waitGroup
+	wg := sync.WaitGroup{}
+	wg.Add(numWorkers)
+
 	// spawn maximum_projects_poller_workers to process project polling in parallel
 	for w := 0; w < numWorkers; w++ {
-		go func() {
-			for {
+		go func(wg *sync.WaitGroup) {
+			defer wg.Done()
+			for p := range projectsToPoll {
 				select {
 				case <-until:
 					return
-				case p := <-projectsToPoll:
-					if e := c.pollProject(p); e != nil {
-						errorStream <- e
-					}
+				case errorStream <- doing(p):
 				}
 			}
-		}()
+		}(&wg)
 	}
-	// process errors coming from pollProject
+	// close the error channel when the workers won't write to it anymore
 	go func() {
-		for ex := range errorStream {
-			errs = append(errs, ex)
-		}
+		wg.Wait()
+		close(errorStream)
 	}()
 	// start processing all the projects configured for this run;
 	// since the channel is buffered because we already know the length of the projects to process,
 	// we can close immediately and the runtime will handle the channel close only when the messages are dispatched
-	for _, pr := range projects {
+	for _, pr := range on {
 		projectsToPoll <- pr
 	}
 	close(projectsToPoll)
-	return errs
+	return errorStream
 }
 
 func (c *Client) pollProjectRefsFromPipelines(projectID, limit int) ([]string, error) {

--- a/cmd/gitlab_test.go
+++ b/cmd/gitlab_test.go
@@ -227,3 +227,22 @@ func pollingResult(until <-chan struct{}, projects <-chan Project, client *Clien
 	}
 	return numErrs
 }
+
+func TestClient_pollProjectsWith(t *testing.T) {
+	c := Client{}
+	message := "some error"
+	doing := func(Project) error {
+		return fmt.Errorf(message)
+	}
+	testProjects := []Project{{Name: "test"}, {Name: "test2"}, {Name: "test4"}, {Name: "test4"}}
+	until := make(chan struct{})
+	errCh := c.pollProjectsWith(4, doing, until, testProjects...)
+	var errCount int
+	for err := range errCh {
+		if assert.Error(t, err) {
+			assert.Equal(t, err.Error(), message)
+			errCount++
+		}
+	}
+	assert.Equal(t, len(testProjects), errCount)
+}

